contracts/contracts/gatekeepers/zupass/ZupassGatekeeper.sol

```diff
@@ -0,0 +1,87 @@
+contract ZupassGatekeeper is SignUpGatekeeper, Ownable(msg.sender) {
+  /// @notice the Zupass event UUID converted to bigint
+  uint256 private immutable validEventId;
+
+  /// @notice the Zupass event signer converted to bigint
+  uint256 private immutable validSigner1;
+  uint256 private immutable validSigner2;
+
+  /// @notice the Groth16Verifier contract address
+  Groth16Verifier private immutable verifier;
+
+  /// @notice the reference to the MACI contract
+  address public maci;
+
+  /// @notice a mapping of ticket IDs to whether they have been used
+  mapping(uint256 => bool) public registeredTickets;
+
+  /// @notice custom errors
+  error AlreadyRegistered();
+  error InvalidProof();
+  error InvalidEventId();
+  error InvalidSigners();
+  error InvalidWatermark();
+  error OnlyMACI();
+  error ZeroAddress();
+
+  /// @notice Creates a new ZupassGatekeeper
+  /// @param _validEventId Zupass event UUID converted to bigint
+  /// @param _validSigner1 Zupass event signer[0] converted to bigint
+  /// @param _validSigner2 Zupass event signer[1] converted to bigint
+  /// @param _verifier The Groth16Verifier contract address
+  constructor(uint256 _validEventId, uint256 _validSigner1, uint256 _validSigner2, address _verifier) payable {
+    validEventId = _validEventId;
+    validSigner1 = _validSigner1;
+    validSigner2 = _validSigner2;
+    verifier = Groth16Verifier(_verifier);
+  }
+
+  /// @notice Adds an uninitialised MACI instance to allow for token signups
+  /// @param _maci The MACI contract interface to be stored
+  function setMaciInstance(address _maci) public override onlyOwner {
+    if (_maci == address(0)) revert ZeroAddress();
+    maci = _maci;
+  }
+
+  /// @notice Registers the user only if they have the Zupass event ticket
+  /// @param _user The user's Ethereum address.
+  /// @param _data The ABI-encoded proof and public signals.
+  function register(address _user, bytes memory _data) public override {
+    if (maci != msg.sender) revert OnlyMACI();
+
+    // Decode the given _data bytes
+    (uint256[2] memory _pA, uint256[2][2] memory _pB, uint256[2] memory _pC, uint256[38] memory _pubSignals) = abi
+      .decode(_data, (uint256[2], uint256[2][2], uint256[2], uint256[38]));
+
+    // Verify proof
+    if (!verifier.verifyProof(_pA, _pB, _pC, _pubSignals)) revert InvalidProof();
+
+    // Event id is stored at index 15
+    if (_pubSignals[15] != validEventId) revert InvalidEventId();
+
+    // Signers are stored at index 13 and 14
+    if (_pubSignals[13] != validSigner1 || _pubSignals[14] != validSigner2) revert InvalidSigners();
+
+    // Watermark is stored at index 37
+    // user address converted to bigint is used as the watermark
+    if (_pubSignals[37] != uint256(uint160(_user))) revert InvalidWatermark();
+
+    // Ticket ID is stored at index 0
+    uint256 ticketId = _pubSignals[0];
+    if (registeredTickets[ticketId]) revert AlreadyRegistered();
+
+    registeredTickets[ticketId] = true;
```

Comment:
state changes should happen before external calls are made (so these 3 lines can be moved before proof verification)